### PR TITLE
Migrate `crate-hash-rustc-version` to `rmake`

### PR DESCRIPTION
**Pull request by rejyr**
Part of #121876.

r? @jieyouxu

try-job: x86_64-gnu-llvm-18
try-job: dist-x86_64-linux

### CODE REVIEW

File: tests/run-make/crate-hash-rustc-version/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    // Now re-compile a.rs with another rustc version
+    rustc()
+        .env("RUSTC_FORCE_RUSTC_VERSION", "deadfeed")
+        .input("a.rs")
+        .crate_type("dylib")
+        .args(&flags)
+        .run();
+
+    // After compiling with a different rustc version, write symbols to disk again.
+    let symbols_after = cmd("nm").args(nm_flag).arg(&dynamic_lib_name("a")).run().stdout_utf8();
```
Comment (jieyouxu) [inline tests/run-make/crate-hash-rustc-version/rmake.rs:37]: Nit: 
```suggestion
    let symbols_after = cmd("nm").args(nm_flag).arg(dynamic_lib_name("a")).run().stdout_utf8();
```

**Review by jieyouxu — APPROVED** [review]
Thanks, looks good in general, just some minor nits. I'll run your PR through a few try-jobs to check.

File: tests/run-make/crate-hash-rustc-version/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+use run_make_support::{cmd, diff, dynamic_lib_name, is_darwin, run, run_fail, rustc};
+
+fn main() {
+    let flags = ["-Cprefer-dynamic", "-Csymbol-mangling-version=v0"];
+    let nm_flag = if is_darwin() { ["-D"].as_slice() } else { [].as_slice() };
+
+    // a.rs is a dylib
+    rustc().input("a.rs").crate_type("dylib").args(&flags).run();
+
+    // Write symbols to disk.
```
[the remark this answers is not in the document]
Comment (Rejyr) [reply tests/run-make/crate-hash-rustc-version/rmake.rs:15]: Writing to disk was a part of the `Makefile` implementation. I'll update the comment to align with the `rmake` implementation.

File: tests/run-make/crate-hash-rustc-version/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+fn main() {
+    let flags = ["-Cprefer-dynamic", "-Csymbol-mangling-version=v0"];
+    let nm_flag = if is_darwin() { ["-D"].as_slice() } else { [].as_slice() };
+
+    // a.rs is compiled to a dylib
+    rustc().input("a.rs").crate_type("dylib").args(&flags).run();
+
+    // Store symbols
+    let symbols_before = cmd("nm").args(nm_flag).arg(dynamic_lib_name("a")).run().stdout_utf8();
```
Comment (Oneirical) [inline tests/run-make/crate-hash-rustc-version/rmake.rs:16]: I've been trying to see if it's possible to remove dependency on `nm` throughout the test suite. Can you check if this works with `llvm_readobj()`? I believe you'd do it like this:

`llvm_readobj().arg("--symbols").input(dynamic_lib_name("a")).run().stdout_utf8()`

File: tests/run-make/crate-hash-rustc-version/rmake.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    // Now re-compile a.rs with another rustc version
+    rustc()
+        .env("RUSTC_FORCE_RUSTC_VERSION", "deadfeed")
+        .input("a.rs")
+        .crate_type("dylib")
+        .args(&flags)
+        .run();
+
+    // After compiling with a different rustc version, store symbols again.
+    let symbols_after = cmd("nm").args(nm_flag).arg(dynamic_lib_name("a")).run().stdout_utf8();
```
Comment (Oneirical) [inline tests/run-make/crate-hash-rustc-version/rmake.rs:37]: Same thing here with `llvm_readobj()`.

File: tests/run-make/crate-hash-rustc-version/Makefile [hunk 5]
```diff
@@ -1,38 +0,0 @@
...
-  NM_CMD := nm -D
-endif
-ifeq ($(UNAME),Darwin)
-  EXT=".dylib"
-  NM_CMD := nm
-endif
-
-ifndef NM_CMD
-all:
-	exit 0
```
Comment (Rejyr) [inline tests/run-make/crate-hash-rustc-version/Makefile:20]: The test is only supposed to run on linux or apple. Is there something like 
```
//@ only-linux
//@ only-apple 
```
that would work?
I think that ^ always ignores, although I haven't tested it.

Comment (jieyouxu) [reply tests/run-make/crate-hash-rustc-version/Makefile:20]: We could try ignore-windows

Comment (jieyouxu) [reply tests/run-make/crate-hash-rustc-version/Makefile:20]: I'll have a look later to i see if theres something like only-unix

File: tests/run-make/crate-hash-rustc-version/rmake.rs [hunk 6]
```diff
@@ -0,0 +1,57 @@
+// Ensure that crates compiled with different rustc versions cannot
+// be dynamically linked.
+
+//@ ignore-cross-compile
+//@ ignore-windows
```
Comment (jieyouxu) [inline tests/run-make/crate-hash-rustc-version/rmake.rs:5]: Suggestion: yeah, so we do have `only-unix`, let's maybe use that?

**Comment by Rejyr** [comment]
@rustbot review

Also needs bors try.

**Review by jieyouxu — APPROVED** [review]
Thank you!

**Pull request merged**